### PR TITLE
Introduce a fail-safe mechanism for the error recorvery of the parser

### LINKED ISSUE #24148: Parser never returns for an incomplete code
**Description:**
Consider following code snippet. This causes parser to infinitely spin cycles.

```ballerina
function foo(){
	, false) : foo(1, false);
}
```

**Affected Versions:**
2.0.0-SNAPSHOT

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
Parser error recovery can cause Infinite loops due to invalid logics of error recovery. This PR introduces an initial approach fail-safe mechanism for such cases until those invalid logics are fixed.

Fixes #24148

## Approach
Limit the number of times parser tries to recover staying on the same token index.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java [hunk 1]
```diff
@@ -82,10 +92,18 @@ public Solution recover(ParserRuleContext currentCtx, STToken nextToken, Object.
...
+        } else {
+            itterCount = 0;
+            previousIndex = currentIndex;
+        }
+
         Result bestMatch = seekMatch(currentCtx);
         if (bestMatch.matches > 0) {
             Solution sol = bestMatch.solution;
-            if (sol != null) {
+            if (sol != null && itterCount < ITTER_LIMIT) {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java:106]: Can we move this check to before calling `seekMatch()`?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java [hunk 2]
```diff
@@ -35,15 +35,25 @@
...
     /**
      * Limit for the distance to travel, to determine a successful lookahead.
      */
     protected static final int LOOKAHEAD_LIMIT = 5;
 
+    /**
+     * Limit for the number of times parser tries to recover staying on the same token index.
+     * This will prevent parser going to infinite loops.
+     */
+    private static final int ITTER_LIMIT = 10;
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java:50]: How about a limit of 5? will it break things?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/incremental/HybridNodeStorage.java [hunk 3]
```diff
@@ -178,4 +178,8 @@ private TextEditRange markAffectedRange(ModulePartNode oldTree, TextEdit textEdi
                 (textRange.startOffset() - affectedTokenStartOffset);
         return new TextEditRange(affectedTokenStartOffset, textRange.endOffset(), newTextLength);
     }
+
+    public int getCurrentIndex() {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/incremental/HybridNodeStorage.java:182]: Looks like other methods don't have the public qualifier. Shall we do the same?
Also, let move this to the other non-private methods section.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java (hunk 2 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java:50]: Yes 5 break things. Trying 7

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java (hunk 1 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java:106]: okay

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/incremental/HybridNodeStorage.java (hunk 3 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/incremental/HybridNodeStorage.java]: Done

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java [hunk 4]
```diff
@@ -35,15 +35,25 @@
     protected final AbstractTokenReader tokenReader;
     protected final BallerinaParserErrorListener errorListener;
     private ArrayDeque<ParserRuleContext> ctxStack = new ArrayDeque<>();
+    private int previousIndex;
```
Comment (sameerajayasoma) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java:38]: `previousIndex` is also not a descriptive name

**Comment by SupunS** [comment]
@rdulmina 
I have sent a PR https://github.com/ballerina-platform/ballerina-lang/pull/24299 resolving the conflicts. 

Let's address the comments in a separate PR.

**Pull request merged**